text/0000-service-hook.md

```diff
@@ -0,0 +1,175 @@
+  token,
+  children,
+}: Props<TService>) {
+  return (
+    <ServiceContext.Consumer>
+      {(getService: (Token<TService>) => TService) => {
+        const provides = getService(token);
+        return children(provides);
+      }}
+    </ServiceContext.Consumer>
+  );
+}
+
+export {ServiceContext};
+
+// components/Example.js
+export default function Example() {
+  return (
+    <ServiceConsumer token={ExampleToken}>
+      {service => (
+        <button onClick={service}>Invoke service</button>
+      )}
+    </ServiceConsumer>
+  );
+}
+```
+
+### Fusion Context
+
+Many plugins depend on the Fusion middleware context. Therefore in addition to getting a service from a hook, we need to allow access to context in a similar fashion. This is done simply enough with Context.
+
+```javascript
+// fusion-react
+export const FusionContext = React.createContext<any>({});
+
+// components/Example.js
+import {useContext} from 'react';
+import {FusionContext} from 'fusion-react';
+
+export default function Example() {
+  // ...
+  const ctx = useContext(FusionContext)
+  // ...
+}
+```
+
+### Context.Provider
+
+The ServiceContext Provider would supply a function that can get the registered service.
+
+The most obvious approach is to simply provide access to the `getService` method on the app instance. This could be done with a very simple plugin.
+
+```javascript
+// fusion-react
+export default class App extends FusionApp {
+  constructor() {
+    // ...
+    this.register(serviceContextPlugin(this));
+  }
+}
+
+function serviceContextPlugin(
+  app: FusionApp
+): FusionPlugin<void, void> {
+  return createPlugin({
+    middleware(): Middleware {
+      return (ctx, next) => {
+        ctx.element = ctx.element && (
+          <FusionContext.Provider value={ctx}>
+            <ServiceContext.Provider value={app.getService}>
+              {ctx.element}
+            </ServiceContext.Provider>
+          </FusionContext.Provider>
+        );
+        return next();
+      };
+    }
+  });
+}
+```
+
+`getService` provides access to the resolved service, which is exactly what we want. Though I can find examples of its usage, `getService` is an undocumented method so it's possibly intended to be private.
+
+FusionContext gives access to the Fusion middleware context object.
+
+### Error handling
+
+Any access to `app.getService` (through `useService` or `ServiceConsumer`) will throw an exception in cases where a token hasn't been registered, the plugin has no `provides`, or the plugin `provides` returns `undefined`. Throwing in these cases shouldn't be detrimental to developer experience since the entire point of Context is to get access to an expected functionality. We have no current method of detecting the difference between these 3 cases. Should we need to support unregisted services, such as for optional plugins, we could potentially add an `app.hasService` method on the base app in `fusion-core` to see if a plugin has been registered.
```

Comment:
Instead of `app.hasService`, could we leverage the `optional` property on tokens for optional semantics?

```js
const service = useService(ExampleToken.optional);
if (service) {
  // token was registered
} else {
  // was not registered, no error thrown
}
```

Reply:
Yeah this is definitely possible and straightforward. This means the exception wouldn't be thrown if the token was optional. Seems like a good idea to have a way to do this, especially since hooks have the shortcoming of being unconditional.

Reply:
Yes! This is a great idea